[DBZ-1214] Ensure PG connection keep alive

addisonj opened the pull request:
This adds a small change to the postgres connector which
works around a problem when large databases take more
than a minute to refresh the schema

This works by creating a simple background thread that ensures we send
statusUpdate messages periodically

I considered if we could refresh the schema before we create the
replication connection but I believe we choose our offset in the log
and then get the schema in order to ensure we don't get out of sync with
any potential schema changes

addisonj commented:
Hrm... this doesn't seem to be fixing the issue, but it might just be that my thread pool is starved, will update with results

gunnarmorling commented:
Interesting; the connector uses the "database.tcpKeepAlive" option by default to avoid this kind of issue. Could you have this disabled by any chance?

addisonj commented:
I haven't disabled tcpKeepAlive, initially, I assumed it wouldn't help as I was assuming that postgres needed not just any tcp  message, but instead a specific standbyStatus message, which serves as the heartbeat. I could be wrong, but I had the impression that postgres will timeout the walsender process if it doesn't get one of those messages, but perhaps it is any tcp activity.

However, upon looking further, I don't actually see where the `TCP_KEEPALIVE` config value is ever read except in a test. It seems to me like the value has no effect, but I am going to confirm that by looking at an older commit.

As far as why this isn't working, it tries to send the message, but it it is never sent, so I am trying to figure that out as well. Upon thinking about this further, I think you could delay opening the socket, as all you need to do is just choose the LSN to start from, but you don't actively have to hold the replication slot. There is a small problem in that some other process could start and take the slot while we fetch metadata, but that would seem to be a problem external to debezium

addisonj commented:
Okay, so more research:

It seems like at the very least, the default of the tcpKeepAlive  being set to true is not added to the Postgres JDBC properties bag (or at least, it wasn't present in the debug logs) and setting it explicitly didn't fix the problem (but I actually didn't ensure that the value was being set).

What was missing in this patch was that the statusUpdateInterval was null, so it wasn't looping as expected, but upon specifically setting that (which is strange, as once again, it should have a default) that got this code to behave properly and it fixed the problem.

I will double check and see if TCP keep alive simply isn't being passed

gunnarmorling commented:
@addisonj The `database.tcpKeepAlive` should be propagated to the connection automatically by virtue of being prefixed with "database." (all those props will be passed to the driver without the prefix). If that's actually not the case, then something is broken IMO. If it actually is passed but doesn't help, then your proposed fix seems like a good approach.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java:
```diff
@@ -127,6 +130,9 @@ protected synchronized void start(BlockingConsumer<ChangeEvent> eventConsumer, C
         ReplicationStream stream = this.replicationStream.get();
+        // once we are streaming changes, we stop the keep alive, as the read loop
+        // will ensure that happens
+        stream.stopKeepAlive();
```
Will this be invoked if the connector stops early (i.e. is shut down before the snapshot is complete)? We must make sure no threads are leaked in such case.

gunnarmorling replied:
@addisonj, I still think there's a (slight) risk of leaking that keep-alive thread under unfortunate circumstances. It seems that `streamChanges()` may be interrupted (by calling `stop()`), before the call flow gets here. I think to be safe `stop()` itself should call `stream.stopKeepAlive()` (which would have to be idempotent for that). WDYT?

@jpechane, how do you feel about this change, does it interfere much with your ongoing work on the connector unification?

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:
```diff
@@ -304,6 +306,34 @@ public Long lastReceivedLsn() {
+                        while (!Thread.currentThread().isInterrupted()) {
+                            try {
+                                stream.forceUpdateStatus();
+                                Thread.sleep(statusUpdateIntervalMillis);
```
Better use `Metronom#park()`.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:
```diff
@@ -304,6 +306,34 @@ public Long lastReceivedLsn() {
+                                Thread.sleep(statusUpdateIntervalMillis);
+                            }
+                            catch (Exception exp) {
+                                throw new RuntimeException("received unexpected exception will perform keep alive", exp);
```
s/perform/performing/?

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:
```diff
@@ -304,6 +306,34 @@ public Long lastReceivedLsn() {
+                if (keepAliveExecutor == null) {
+                    keepAliveExecutor = service;
+                    keepAliveExecutor.submit(() -> {
+                        while (!Thread.currentThread().isInterrupted()) {
```
Reading the docs of `shutdownNow()` it seems it's not guaranteed that it actually will interrupt the running thread. I think a safer way would be to have a `AtomicBoolean` or so as an instance field in this class and use this as the stopping signal.

addisonj commented:
@gunnarmorling getting back to this, yes, it is passed in the properties as expected, but doesn't appear to help anything. Have confirmed it does all work with this patch, so I think it is expecting the explicit standby message rather than just the keep alive

Pushed up an updated patch, addressed your comments and also sneaking in a really small logging change for a very chatty log I noticed.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:
```diff
@@ -304,6 +312,40 @@ public Long lastReceivedLsn() {
+                if (keepAliveExecutor == null && statusUpdateIntervalMillis > 0) {
+                    LOGGER.info("starting keep alive thread");
+                    keepAliveExecutor = service;
+                    Metronome metronome = Metronome.sleeper(Duration.ofMillis(statusUpdateIntervalMillis), Clock.SYSTEM);
```
Can you make `statusUpdateIntervalMillis` a `Duration` itself (right up to the point where the value is obtained from the config); we try to converge to the more semantic temporal types globally, so this is a good occassion to continue that effort.

gunnarmorling commented:
*ping* Just to make sure you saw the latest comments, @addisonj :)

addisonj commented:
Mentioned in other issue, but yes, have seen this just have been diverted
to something else for a bit but will try and get some changes in by Friday.

On Tue, Apr 23, 2019 at 2:16 AM Gunnar Morling <notifications@github.com>
wrote:

> *ping* Just to make sure you saw the latest comments, @addisonj
> <https://github.com/addisonj> :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/debezium/debezium/pull/835#issuecomment-485692263>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAE3LAYBHHH223UWDPXYEMLPR3AVTANCNFSM4HDXK45Q>
> .
>

addisonj commented:
@gunnarmorling okay, i updated those few things, let me know if there is anything else here. Will work on the other PR tomorrow and see if I can figure out a good answer for all of that

gunnarmorling commented:
Thanks, @addisonj! I've pushed one more commit, most notably for using a flag to signal the thread shutdown as suggested above. With that I think it's good to go. Thanks a lot!

gunnarmorling commented:
@jpechane, can you check out my additional commit and merge if you don't object? Thanks!

jpechane commented:
@gunnarmorling The testsuite is failing now with NPE for many tests

gunnarmorling commented:
Hum, which connector? PG is fine.

>

jpechane commented:
@gunnarmorling Both postgres plugins

gunnarmorling commented:
Where are you seeing this? Do you have a reference?

jpechane commented:
https://travis-ci.org/debezium/debezium/builds/527167307?utm_source=github_status&utm_medium=notification

gunnarmorling commented:
Agh, got it now. I thought it was merged already, so I was trying with master ;)

Looking into it.

gunnarmorling commented:
@jpechane, pushed one more commit to fix it.

jpechane commented:
Applied, thanks @addisonj!
